src/lsm/compaction.zig

```diff
@@ -819,174 +886,136 @@ pub fn CompactionType(
             }
         }
 
-        /// Plan the work for the beat:
-        /// - determine our quota in terms of minimum number of input values to process,
-        /// - reserve grid blocks for the output.
         pub fn beat_commence(
             compaction: *Compaction,
-            beats_remaining: u64,
+            values_count: u64,
         ) void {
             assert(compaction.idle());
+            assert(compaction.pool.?.idle());
             assert(compaction.stage == .paused);
-            if (compaction.move_table) assert(compaction.quotas.bar_done());
-            assert(compaction.quotas.done >= compaction.quotas.beat);
-            defer assert(compaction.quotas.done <= compaction.quotas.beat);
-
-            assert(beats_remaining > 0);
-            assert(beats_remaining <= half_bar_beat_count);
-
-            // Calculate how many values we have to compact each beat, to self-correct our pacing.
-            // Pacing will have imperfections due to rounding up to fill target value blocks and
-            // immutable table filtering duplicate values.
-            const values_remaining = compaction.quotas.bar - compaction.quotas.done;
-            const beat_values_quota = stdx.div_ceil(values_remaining, beats_remaining);
-            compaction.quotas.beat = compaction.quotas.done + beat_values_quota;
-            assert(compaction.quotas.done <= compaction.quotas.beat);
+            assert(compaction.block_queues_empty_output());
+            // We may be carrying over some blocks from the previous beat.
+            maybe(compaction.block_queues_empty_input());
+
+            if (compaction.move_table) assert(compaction.quotas.bar_exhausted());
+
+            // Run the compaction up to completion of the bar quota, if possible.
+            const values_remaining = (compaction.quotas.bar - compaction.quotas.bar_done);
+            const quota_beat = @min(values_count, values_remaining);
```

Comment:
```suggestion
            compaction.quotas.beat = @min(values_count, values_remaining);
```

to avoid introducing an alias